src/controller.ts

```diff
@@ -74,59 +87,60 @@ export class EthrDidController {
-    return result[0]
+    return this.contract.identityOwner(address, { blockTag })
   }
 
-  async attachContract(controller?: address | Promise<address>): Promise<Contract> {
-    const currentOwner = controller ? await controller : await this.getOwner(this.address, 'latest')
-    const signer = this.signer
-      ? this.signer
-      : (<JsonRpcProvider>this.contract.provider).getSigner(currentOwner) || this.contract.signer
-    return this.contract.connect(signer)
+  async attachContract(controller?: AddressLike): Promise<Contract> {
+    let currentOwner = controller ? await controller : await this.getOwner(this.address, 'latest')
+    if (typeof currentOwner !== 'string') currentOwner = await (controller as Addressable).getAddress()
+    let signer
+    if (this.signer) {
+      signer = this.signer
+    } else {
+      if (!this.contract) throw new Error(`No contract configured`)
+      if (!this.contract.runner) throw new Error(`No runner configured for contract`)
+      if (!this.contract.runner.provider) throw new Error(`No provider configured for runner in contract`)
+      signer = (await (<JsonRpcProvider>this.contract.runner.provider).getSigner(currentOwner)) || this.contract.signer
+    }
+    return this.contract.connect(signer) as Contract // Needed because ethers attach returns a BaseContract
   }
 
-  async changeOwner(newOwner: address, options: CallOverrides = {}): Promise<TransactionReceipt> {
+  async changeOwner(newOwner: address, options: Overrides = {}): Promise<TransactionReceipt> {
     // console.log(`changing owner for ${oldOwner} on registry at ${registryContract.address}`)
     const overrides = {
       gasLimit: 123456,
       ...options,
-    }
-
-    const contract = await this.attachContract(overrides.from)
+    } as Overrides
+    const contract = await this.attachContract(overrides.from ?? undefined)
     delete overrides.from
 
-    const ownerChange = await contract.functions.changeOwner(this.address, newOwner, overrides)
+    const ownerChange = await contract.changeOwner(this.address, newOwner, overrides)
     return await ownerChange.wait()
   }
 
   async createChangeOwnerHash(newOwner: address) {
     const paddedNonce = await this.getPaddedNonceCompatibility()
 
-    const dataToHash = hexConcat([
+    const dataToHash = concat([
       MESSAGE_PREFIX,
-      this.contract.address,
+      await this.contract.getAddress(),
       paddedNonce,
       this.address,
-      concat([toUtf8Bytes('changeOwner'), newOwner]),
+      getBytes(concat([toUtf8Bytes('changeOwner'), newOwner])),
     ])
     return keccak256(dataToHash)
   }
 
-  async changeOwnerSigned(
-    newOwner: address,
-    metaSignature: MetaSignature,
-    options: CallOverrides = {}
-  ): Promise<TransactionReceipt> {
+  async changeOwnerSigned(newOwner: address, metaSignature: MetaSignature, options = {}): Promise<TransactionReceipt> {
     const overrides = {
       gasLimit: 123456,
       ...options,
-    }
+    } as any
```

Comment:
Please avoid using `any` here. This should be an object that can override the transaction parameters.

Same for all the other similar code blocks that are cast `as any`